elfcore/src/coredump.rs

```diff
@@ -1163,7 +1182,68 @@ fn write_elf_note<T: Write>(
 
     writer.write_all(data)?;
     written += data.len();
-    written += writer.align_position(ELF_NOTE_PADDING as u64)?;
+    written += writer.align_position(ELF_NOTE_ALIGN)?;
+
+    Ok(written)
+}
+
+fn write_elf_note_file<T: Write>(
+    writer: &mut ElfCoreWriter<T>,
+    note_kind: u32,
+    name_bytes: &[u8],
+    file: &mut File,
+    note_len: usize,
+) -> Result<usize, CoreError> {
+    let mut written = 0_usize;
+
+    let header_and_name =
+        std::mem::size_of::<Elf64_Nhdr>() + round_up(name_bytes.len() + 1, ELF_NOTE_ALIGN);
+    let data_len = note_len - header_and_name;
+    written += write_elf_note_header(writer, note_kind, name_bytes, data_len)?;
+
+    tracing::debug!(
+        "Writing note payload {} bytes at offset {}...",
+        data_len,
+        writer.stream_position()?
+    );
+
+    let max_len = data_len - std::mem::size_of::<u32>();
+    let mut buf = [0_u8; BUFFER_SIZE];
+    let mut total = 0;
+    loop {
+        match file.read(&mut buf) {
+            // if eof or would block, we are done
+            Ok(0) => break,
+            Err(ref err) if err.kind() == ErrorKind::WouldBlock => break,
```

Comment:
You've done this here because you're dumping /dev/kmsg specifically. But this doesn't seem appropriate in a generic crate.

Reply:
Instead of taking a `File`, can't we take a `&mut dyn Read` or something?

Reply:
would this not work for a normal file? if you pass in a file that hasn't been opened with the nonblock flag, then it shouldn't return the WouldBlock error right, it would just block?

Reply:
I went ahead and switched from File to Box<dyn Read> implemented in the called (see that code as well)